### PR TITLE
feat: Collect staking rewards directly from custody contract 

### PR DESCRIPTION
**Pull request by mandrean**
- Collect staking rewards directly from custody contract
- Track total cumulative rewards
- Use cw-plus-storage

and some minor fixes/tweaks.

### CODE REVIEW

File: contracts/custody_beth/src/contract.rs [hunk 1]
```diff
@@ -84,6 +84,7 @@ pub fn execute(
             let borrower_addr = deps.api.addr_validate(&borrower)?;
             liquidate_collateral(deps, info, liquidator_addr, borrower_addr, amount)
         }
+        ExecuteMsg::WithdrawStakingRewards {} => Err(ContractError::Unauthorized {}),
```
Comment (alwin-peng) [inline contracts/custody_beth/src/contract.rs:87]: beth needs the same changes as bluna

**Comment by alwin-peng** [comment]
besides beth comment, LGTM

**Review by MSNTCS — CHANGES_REQUESTED** [review]

File: contracts/custody_bluna/src/state.rs [hunk 2]
```diff
@@ -106,3 +106,77 @@ fn calc_range_start(start_after: Option<CanonicalAddr>) -> Option<Vec<u8>> {
         v
     })
 }
+
+// rewards / collateral
+const KEY_GLOBAL_INDEX: &[u8] = b"global_index";
```
[the remark this answers is not in the document]
Comment (mandrean) [reply contracts/custody_bluna/src/state.rs:111]: >Please add these states to the beginning of the file.

I honestly think `package-by-feature` is superior to `package-by-layer`, but sure, can do.

File: contracts/custody_bluna/src/state.rs [hunk 3]
```diff
@@ -106,3 +106,77 @@ fn calc_range_start(start_after: Option<CanonicalAddr>) -> Option<Vec<u8>> {
...
+const KEY_USER_REWARDS: &[u8] = b"user_rewards";
```
[the remark this answers is not in the document]
Comment (mandrean) [reply contracts/custody_bluna/src/state.rs:112]: I thought about it but noticed we don't even use `cw-plus` anywhere in the repo. Sounds good!

File: contracts/custody_beth/src/contract.rs (hunk 1 above)
Comment (mandrean) [reply contracts/custody_beth/src/contract.rs]: Ah ok. Done 👍

File: contracts/custody_bluna/src/state.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mandrean) [reply contracts/custody_bluna/src/state.rs]: >please also merge KEY_GLOBAL_INDEX and KEY_TOTAL_CUMULATIVE_REWARDS in one single struct.

Done!

File: contracts/custody_bluna/src/state.rs [hunk 4]
```diff
@@ -106,3 +106,77 @@ fn calc_range_start(start_after: Option<CanonicalAddr>) -> Option<Vec<u8>> {
...
+
+pub fn save_total_cumulative_rewards(
+    storage: &mut dyn Storage,
+    new_rewards: &Uint256,
+) -> StdResult<()> {
+    let mut rewards = Singleton::new(storage, KEY_TOTAL_CUMULATIVE_REWARDS);
+    rewards.save(new_rewards)
+}
+
+pub fn read_total_cumulative_rewards(storage: &dyn Storage) -> Uint256 {
```
[the remark this answers is not in the document]
Comment (mandrean) [reply contracts/custody_bluna/src/state.rs]: Done!

File: contracts/custody_bluna/src/state.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mandrean) [reply contracts/custody_bluna/src/state.rs]: Done!

**Comment by mandrean** [comment]
[Users collect staking rewards directly from custody contract](https://trello.com/c/dGNGoFIU/23-users-collect-staking-rewards-directly-from-custody-contract)

**Pull request merged**